charmcraft/extensions/registry.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Extension registry."""
+
+from typing import TYPE_CHECKING, Dict, List, Type
+
+from charmcraft import errors
+
+if TYPE_CHECKING:
+    from charmcraft.extensions import Extension
+
+    ExtensionType = Type[Extension]
+
+_EXTENSIONS: Dict[str, "ExtensionType"] = {}
+
+
+def get_extension_names() -> List[str]:
+    """Obtain a extension class given the name.
+
+    :param name: The extension name.
+    :return: The list of available extensions.
+    :raises ExtensionError: If the extension name is invalid.
+    """
+    return list(_EXTENSIONS.keys())
+
+
+def get_extension_class(extension_name: str) -> "ExtensionType":
+    """Obtain a extension class given the name.
+
+    :param name: The extension name.
+    :return: The extension class.
+    :raises ExtensionError: If the extension name is invalid.
+    """
+    try:
+        return _EXTENSIONS[extension_name]
+    except KeyError as key_error:
+        raise errors.ExtensionError(f"Extension {extension_name!r} does not exist") from key_error
```

Comment:
aren't tracebacks more thorough though?

Reply:
If the traceback adds useful information, definitely. But I think this is one of the rare cases where it doesn't. This snippet is functionally equivalent to

```python
if extension_name in _EXTENSIONS:
    return _EXTENSIONS[extension_name]
raise errors.ExtensionError(...)
```

and I doubt we'd ask for more context if it were written using `if` rather than `try`.

It's a minimal difference since it only affects charmcraft with `--verbosity=debug`, but here's an example script:

```python
import sys
import craft_cli

craft_cli.emit.init(craft_cli.EmitterMode.DEBUG, "testcraft", "hi")


def emit_ext_err(extension):
    try:
        {}[extension]
    except KeyError as e:
        if "--no-cause" in sys.argv:
            e = None
        raise craft_cli.CraftError(f"Extension {extension!r} does not exist") from e


try:
    emit_ext_err("invalid_extension")
except craft_cli.CraftError as e:
    craft_cli.emit.error(e)
```

The output in the two cases looks like:

```
(charmcraft) lengau@hyperion:~/Work/Code/charmcraft$ python test.py 
2023-09-07 13:44:12.119 hi                                                                                                                                                                                                                                           
2023-09-07 13:44:12.119 Logging execution to '/home/lengau/.local/state/testcraft/log/testcraft-20230907-134412.119016.log'                                                                                                                                          
2023-09-07 13:44:12.119 Extension 'invalid_extension' does not exist                                                                                                                                                                                                 
2023-09-07 13:44:12.120 Traceback (most recent call last):                                                                                                                                                                                                           
2023-09-07 13:44:12.120   File "/home/lengau/Work/Code/charmcraft/test.py", line 9, in emit_ext_err                                                                                                                                                                  
2023-09-07 13:44:12.120     {}[extension]                                                                                                                                                                                                                            
2023-09-07 13:44:12.120 KeyError: 'invalid_extension'                                                                                                                                                                                                                
2023-09-07 13:44:12.120 Full execution log: '/home/lengau/.local/state/testcraft/log/testcraft-20230907-134412.119016.log'                                                                                                                                           
(charmcraft) lengau@hyperion:~/Work/Code/charmcraft$ python test.py --no-cause
2023-09-07 13:44:16.709 hi                                                                                                                                                                                                                                           
2023-09-07 13:44:16.709 Logging execution to '/home/lengau/.local/state/testcraft/log/testcraft-20230907-134416.708554.log'                                                                                                                                          
2023-09-07 13:44:16.709 Extension 'invalid_extension' does not exist                                                                                                                                                                                                 
2023-09-07 13:44:16.709 Full execution log: '/home/lengau/.local/state/testcraft/log/testcraft-20230907-134416.708554.log' 
```

Thinking through UX now, maybe we should put a `details=f"Registered extensions: {get_extension_names()}"`

Reply:
OK, I changed it to raise from None with details.